src/charm.py

```diff
@@ -100,23 +98,47 @@ def _on_config_changed(self, event):
 
     def publish_config(self):
         """Generate config file and publish to all workers."""
-        mimir_config = self.coordinator.build_config(dict(self.config))
+        mimir_config = self.coordinator.build_config(self._s3_storage_data)
         self.cluster_provider.publish_configs(mimir_config)
 
+    def _s3_connection_info(self):
+        """Parse a _S3StorageBackend object from relation data."""
+        if not self.s3_requirer.relations:
+            self._s3_storage_data = _S3StorageBackend()
+            return
+        raw = self.s3_requirer.get_s3_connection_info()
+        raw["access_key"], raw["secret_key"] = raw.pop("access-key", ""), raw.pop("secret-key", "")
+        self._s3_storage_data = _S3StorageBackend(**raw)
```

Comment:
Since we use Pydantic, you can use aliases to spare you the need of doing this translation shenanigans